Doc Cookbook Auto Interface Implementation (#73906)
dotnet/roslyn: docs/features/incremental-generators.cookbook.md

### Auto interface implementation

**User scenario:** As a generator author I want to be able to implement the properties of interfaces passed as arguments to a decorator of a class automatically for a user

**Solution:** Require the user to decorate the class with the `[AutoImplement]` Attribute and pass as arguments the types of the interfaces they want to self-implement themselves; The classes that implement the attribute have to be `partial class`.
Provide that attribute in a `RegisterPostInitializationOutput` step. Register for callbacks on the classes with
`ForAttributeWithMetadataName` using the fullyQualifiedMetadataName `FullyQualifiedAttributeName`, and use tuples (or create an equatable model) to pass along that information.

> The Attribute could work for Structs too, the example was kept simple on purpose for the workbook sample

Reviewer: ```suggestion
The attribute could work for structs too, the example was kept simple on purpose for the workbook sample.
```

**Example:**

```csharp
public interface IUserInterface
{
    int InterfaceProperty { get; set; }
}

public interface IUserInterface2
{
    float InterfacePropertyOnlyGetter { get; }
}

> ```
> 
> ```csharp

Reviewer: Let's keep these snippets together so it's clear that they're both user code.

```suggestion

```

    public void Initialize(IncrementalGeneratorInitializationContext context)
    {
        context.RegisterPostInitializationOutput(ctx =>
        {
            //Generate the AutoImplementProperties Attribute
            const string autoImplementAttributeDeclarationCode = $$"""
// <auto-generated/>
using System;
namespace {{AttributeNameSpace}};

[AttributeUsage(AttributeTargets.Class, Inherited = false, AllowMultiple = false)]
sealed class {{AttributeClassName}} : Attribute
{
    public Type[] InterfacesTypes { get; }
    public {{AttributeClassName}}(params Type[] interfacesTypes)
    {
        InterfacesTypes = interfacesTypes;
    }
}
""";
            ctx.AddSource($"{AttributeClassName}.g.cs", autoImplementAttributeDeclarationCode);
        });

        IncrementalValuesProvider<ClassModel> provider = context.SyntaxProvider.ForAttributeWithMetadataName(
            fullyQualifiedMetadataName: FullyQualifiedAttributeName,
            predicate: static (node, cancellationToken_) => node is ClassDeclarationSyntax,
            transform: static (ctx, cancellationToken) =>
            {

>                 ISymbol classSymbol = ctx.SemanticModel.GetDeclaredSymbol(ctx.TargetNode, cancellationToken)!;

Reviewer: ```suggestion
                ISymbol classSymbol = ctx.TargetSymbol;
```

        {
            foreach (InterfaceModel interfaceModel in classModel.Interfaces)
            {
                StringBuilder sourceBuilder = new($$"""
                    // <auto-generated/>
                    namespace {{classModel.NameSpace}};

                    public partial class {{classModel.Name}} : {{interfaceModel.FullyQualifiedName}}
                    {

                    """);

                foreach (string property in interfaceModel.Properties)
                {
                    sourceBuilder.AppendLine(property);
                }

                sourceBuilder.AppendLine("""
                    }
                    """);

                //Concat class name and interface name to have unique file name if a class implements two interfaces with AutoImplement Attribute
                string generatedFileName = $"{classModel.Name}_{interfaceModel.FullyQualifiedName}.g.cs";
                context.AddSource(generatedFileName, sourceBuilder.ToString());
            }
        });
    }

    private static EquatableList<InterfaceModel> GetInterfaceModels(ISymbol classSymbol)
    {
        EquatableList<InterfaceModel> ret = [];

>         AttributeData attribute = classSymbol.GetAttributes().First(a => a.AttributeClass!.Name == AttributeClassName);

Reviewer: Just get this from the context. The `Attributes` property will have all the matched attributes; since `AllowMultiple` is false, you can just access the `[0]`th index and not care about any duplicates with errors.

>         INamedTypeSymbol[] interfaceSymbols = attribute.ConstructorArguments[0].Values.Select(x => (x.Value as INamedTypeSymbol)!).ToArray();

Reviewer: You can't just assume that everything bound correctly and is a non-null type. The user could have an error.

```suggestion
        // A separate analyzer should error on using a non-interface type in this list.
        INamedTypeSymbol[] interfaceSymbols = attribute.ConstructorArguments[0].Values.Select(x => x.Type).Where(t => t is { TypeKind: TypeKind.Interface }).ToArray()!;
```

Author: You are right about "You can't just assume that everything bound correctly and is a non-null type. The user could have an error.", but it seems your solution is not working.
1) Every element of "_**attribute.ConstructorArguments[0].Values**_" in this case has _**Type.TypeKind == TypeKind.Class**_
2) "_**attribute.ConstructorArguments[0].Values.Select(x => x.Type).Where(t => t is { TypeKind: TypeKind.Interface }).ToArray()!**_" is a "_**ITypeSymbol[]**_", not a "_**INamedTypeSymbol[]**_". So the subsequent foreach is not working, cause it finds properties whose type is assembly.

I'll work on this

Reviewer: 1. Well, then you're not passing an interface.
2. All interfaces are `INamedTypeSymbol`s, you can just add a `Cast<INamedTypeSymbol>()` and it'll work fine.

Author: Sorry, probably I'm not understanding something
![Screenshot 2024-06-21 213852](https://github.com/dotnet/roslyn/assets/92273914/b93ae2de-2eb8-44ac-bfee-df9fd7f3f626)
As you can see, even if I write "_**[AutoImplementProperties(typeof(IUserInterface), typeof(IUserInterface2))]**_" the TypeKind is "_**Class**_"

And both "_**tempo**_" and "**_tempo2_**" arrays have 61 elements (you can see some of them in the screenshot) instead of the 1 I expected, as every Interface in my sample only have 1 property

I'm keeping updated this repo in order to test every update I'm committing
https://github.com/bdbonazz/AutoImplementGenerator

Reviewer: Ah, I think I see. The select should select the `Value`, not `Type`, and check for `INamedTypeSymbol` there. `Type` in this case is the actual type of the argument, which is `System.Type`, not the type bound by the `typeof`. `Value` has that.

Author: Perfect
I added the check on TypeKind.Interface

### Auto interface implementation

**User scenario:** As a generator author I want to be able to implement the properties of interfaces passed as arguments to a decorator of a class automatically for a user

**Solution:** Require the user to decorate the class with the `[AutoImplement]` Attribute and pass as arguments the types of the interfaces they want to self-implement themselves; The classes that implement the attribute have to be `partial class`.
Provide that attribute in a `RegisterPostInitializationOutput` step. Register for callbacks on the classes with
`ForAttributeWithMetadataName` using the fullyQualifiedMetadataName `FullyQualifiedAttributeName`, and use tuples (or create an equatable model) to pass along that information.
The attribute could work for structs too, the example was kept simple on purpose for the workbook sample.

**Example:**

```csharp
public interface IUserInterface
{
    int InterfaceProperty { get; set; }
}

public interface IUserInterface2
{
    float InterfacePropertyOnlyGetter { get; }

> }

Reviewer: ```suggestion
}

```

### Auto interface implementation

**User scenario:** As a generator author I want to be able to implement the properties of the interfaces decorated with a specific attribute automatically for a user.

**Solution:** Require the user to decorate the interface that they want to self-implement with the `[AutoImplement]` Attribute; Require the user to make the classes that implement the decorated interface be a `partial class`.
Provide that attribute in a `RegisterPostInitializationOutput` step. Register for callbacks on the classes with

> `CreateSyntaxProvider` using the predicate `node is ClassDeclarationSyntax classDeclarationSyntax && classDeclarationSyntax.BaseList is not null && classDeclarationSyntax.BaseList.Types.Count > 0` to collect the classes that possibly implements at least one interface, And use tuples (or create an equatable model) to pass along that information.

Reviewer: Looks like this description still needs to be updated.

**Example:**

```csharp
public interface IUserInterface
{
    public int InterfaceProperty { get; set; }
}

public interface IUserInterface2
{
    public float InterfacePropertyOnlyGetter { get; }
}
```

```csharp
[AutoImplement(nameof(IUserInterface), nameof(IUserInterface2))]
public partial class UserClass
{
    public string UserProp { get; set; }
}
```

```csharp
#nullable enable
[Generator]
public class AutoImplementGenerator : IIncrementalGenerator
{
    private const string NameOfString = "nameof(";
    private const string AutoImplementAttributeNameSpace = "AttributeGenerator";
    private const string AutoImplementAttributeName = "AutoImplement";
    private const string AutoImplementAttributeClassName = $"{AutoImplementAttributeName}Attribute";
    private const string FullyQualifiedMetadataName = $"{AutoImplementAttributeNameSpace}.{AutoImplementAttributeClassName}";

    public void Initialize(IncrementalGeneratorInitializationContext context)
    {
        context.RegisterPostInitializationOutput(ctx =>
        {
            //Generate the AutoImplement Attribute
            const string autoImplementAttributeDeclarationCode = $$"""

> using System;

Reviewer: It's good practice to include the autogenerated comment as well, even when using the `.g.cs` extension, just to make sure it's obvious.

```suggestion
// <auto-generated/>
using System;
```

namespace {{AutoImplementAttributeNameSpace}};

> [AttributeUsage(AttributeTargets.Class, Inherited = false, AllowMultiple = false)]

Reviewer: Structs too? Fine if you want to say that's too complex for the workbook sample, but in that case, it would be good to call out the limitation in a comment explicitly.

sealed class {{AutoImplementAttributeClassName}} : Attribute
{
    public string[] InterfacesNames { get; }
    public {{AutoImplementAttributeClassName}}(params string[] interfacesNames)
    {
        InterfacesNames = interfacesNames;
    }
}
""";
            ctx.AddSource($"{AutoImplementAttributeClassName}.g.cs", autoImplementAttributeDeclarationCode);
        });

        IncrementalValuesProvider<Model> provider = context.SyntaxProvider.ForAttributeWithMetadataName(
            fullyQualifiedMetadataName: FullyQualifiedMetadataName,
            predicate: static (node, cancellationToken_) => node is ClassDeclarationSyntax,
            transform: static (ctx, cancellationToken) =>
            {
                ClassDeclarationSyntax classDeclarationSyntax = (ClassDeclarationSyntax)ctx.TargetNode;
                
                string className = classDeclarationSyntax.Identifier.ValueText;

>                 string classNameSpace = GetClassNameSpace(classDeclarationSyntax.Parent);

Reviewer: Don't do either of these things. The context has the type symbol for the attributed class; get this info directly from the symbol, not from syntax. This code will not work robustly; it won't handle nested types, types in the global namespace, or syntax that deviates from the standard single-namespace form.

                InterfaceModel[] interfaces = GetInterfaceModels(ctx.SemanticModel.Compilation, classDeclarationSyntax);

                return new Model(
                    className,
                    classNameSpace,
                    interfaces
                    );
            }).Where(m => m is not null);

>         context.RegisterSourceOutput(provider.Collect(), Execute);

Reviewer: Don't collect here; just pass in `provider` directly, and remove the outer `foreach` from `Execute`. I'd also rename `Execute` to be something more descriptive, rather than letting it be a relic of `ISourceGenerator`.

    }
    private static string GetClassNameSpace(SyntaxNode? parent)
    {
        return parent is NamespaceDeclarationSyntax namespaceDeclarationSyntax
            ? namespaceDeclarationSyntax.Name.ToString()
            : parent is FileScopedNamespaceDeclarationSyntax fileScopedNamespaceDeclarationSyntax
            ? fileScopedNamespaceDeclarationSyntax.Name.ToString()
            : AutoImplementAttributeNameSpace;
    }

    private static InterfaceModel[] GetInterfaceModels(Compilation compilation, ClassDeclarationSyntax classDeclarationSyntax)
    {
        //Get class usings in order to build full interface name if needed
        string[] usings = GetUsings(classDeclarationSyntax.SyntaxTree.GetRoot());
        string[] interfacesNames = GetInterfacesNames(classDeclarationSyntax.AttributeLists);

        List<InterfaceModel> ret = [];
        foreach (string interfaceName in interfacesNames)
        {
            INamedTypeSymbol? interfaceSymbol = GetInterfaceSymbol(compilation, usings, interfaceName);
            if (interfaceSymbol is null)
                continue;

>             //Get interface usings, in order to be sure to have the needed usings for the properties that will be generated

Reviewer: I would not try to copy usings over. Instead, fully qualify everything. There's no telling what will conflict in the rest of the user's file with your own code. It will also not handle interfaces defined in another assembly.

            string[] interfaceUsings = interfaceSymbol
                .DeclaringSyntaxReferences
                .FirstOrDefault()?
                .GetSyntax()
                .SyntaxTree
                .GetRoot()
                .DescendantNodes()
                .OfType<UsingDirectiveSyntax>()
                .Select(x => x.ToString())
                .ToArray() ?? [];

            IPropertySymbol[] interfaceProperties = interfaceSymbol
                .GetMembers()
                .OfType<IPropertySymbol>()
                .ToArray();

            InterfacePropertyModel[] propertyModels = interfaceProperties
                .Select(interfaceProperty =>
                {
                    /*Using "interfaceProperty.Type" instead of "interfaceProperty.Type.Name" in order to not have error in specific cases.
                    Es. the type "int" has Name "Int32", but writing "Int32" casue compilation error if we are not using "System" namespace*/

>                     string type = interfaceProperty.Type.ToString();

Reviewer: Use an appropriate `ToDisplayString()` instead. The same holds true for everywhere else you are converting type symbols to strings. You likely want `SymbolDisplayFormat.FullyQualifiedFormat`.

                    return new InterfacePropertyModel(type, interfaceProperty.Name, interfaceProperty.SetMethod is not null);
                })
                .ToArray();

            string containingNamespace = interfaceSymbol.ContainingNamespace.ToString();

            ret.Add(new InterfaceModel(interfaceName, containingNamespace, interfaceUsings, propertyModels));
        }

        return [.. ret];
    }
    private static string[] GetUsings(SyntaxNode root)
    {
        return root.DescendantNodes()
                .OfType<UsingDirectiveSyntax>()
                .Select(usingDirective =>
                {
                    //get string "System" from string "using System;"
                    string usingName = usingDirective.ToString().Split(' ').Last().TrimEnd(';');
                    return usingName;
                })
                .ToArray();

>     }

Reviewer: I would not try to copy usings over. Instead, fully qualify everything. There's no telling what will conflict in the rest of the user's file with your own code.

        {
            foreach (AttributeSyntax attribute in attributeList.Attributes)
            {
                if (!attribute.Name.ToString().Equals(AutoImplementAttributeName) || attribute.ArgumentList is null)
                    continue;

                return attribute.ArgumentList.Arguments
                    .Select(x => GetInterfaceName(x.Expression.ToString()))
                    .ToArray();
            }
        }

        return [];
    }
    private static string GetInterfaceName(string expression)
    {
        // Empty String
        if (string.IsNullOrWhiteSpace(expression))
            return expression;

        // Interface Name got by nameof()
        if (expression.StartsWith(NameOfString))
            return expression.Substring(NameOfString.Length, expression.Length - NameOfString.Length - 1);

        //Trimming ""
        string ret = expression.Trim('"');
        return ret;
    }
    private static INamedTypeSymbol? GetInterfaceSymbol(Compilation compilation, string[] nameSpaces, string implementedInterfaceName)
    {
        // Try Get without NameSpace if interface was already Fully Qualified

>         INamedTypeSymbol? interfaceSymbol = compilation.GetTypeByMetadataName(implementedInterfaceName);

Reviewer: You don't need to do this. Change the `AutoImplement` attribute ctor to be `Type[]` instead of `string[]`, and then just leverage the attribute arguments which will already be types. It's both more robust doesn't need the lookup mechanism you're using here, which will likely fail.

        if (interfaceSymbol is not null)
            return interfaceSymbol;

        foreach (string nameSpace in nameSpaces)
        {
            interfaceSymbol = compilation.GetTypeByMetadataName($"{nameSpace}.{implementedInterfaceName}");
            if (interfaceSymbol is not null)
                return interfaceSymbol;
        }

        return null;
    }

    private record Model(string ClassName, string ClassNameSpace, InterfaceModel[] Interfaces);

>     private record InterfaceModel(string Name, string NameSpace, string[] Usings, InterfacePropertyModel[] Properties);

Reviewer: Use an equatable list implementation, rather than arrays.

Author: I'm really grateful for all your suggestions
By 'equatable list implementation' do you mean a custom implementation like 'public class EquatableList<T> : List<T>, IEquatable<EquatableList<T>>' or would you prefer another solution?

Reviewer: That would work. Here's what the cookbook says further up:

> Be careful of collection types in your models. Most built-in collection types in .NET do not do value equality by default. Arrays, ImmutableArray<T> and List<T>, for example, use reference equality, not value equality. We suggest that most generator authors use a wrapper type around arrays to augment them with value-based equality.

    private record InterfacePropertyModel(string Type, string Name, bool HasSetter);

    private void Execute(SourceProductionContext context, ImmutableArray<Model> models)
    {
        foreach (Model model in models)
        {
            foreach (InterfaceModel interfaceModel in model.Interfaces)
            {
                string interfaceUsings = string.Join(string.Empty, interfaceModel.Usings);

                StringBuilder sourceBuilder = new($$"""
                    {{interfaceUsings}}
                    using {{interfaceModel.NameSpace}};

                    namespace {{model.ClassNameSpace}};

                    partial class {{model.ClassName}} : {{interfaceModel.Name}}
                    {

                    """);
                foreach (InterfacePropertyModel property in interfaceModel.Properties)
                {
                    //Check if property has a setter
                    string setter = property.HasSetter
                        ? "set; "
                        : string.Empty;

                    sourceBuilder.AppendLine($$"""

>                             public {{property.Type}} {{property.Name}} { get; {{setter}}}

Reviewer: Since this is only auto-implementing properties as `get; set;`, it would probably be good to narrow the name of `AutoImplementAttribute` to convey that up front for people reading the sample.

        });
    }

    private static EquatableList<InterfaceModel> GetInterfaceModels(AttributeData attribute)
    {
        EquatableList<InterfaceModel> ret = [];

        if (attribute.ConstructorArguments.Length == 0)
            return ret;

        foreach(TypedConstant constructorArgumentValue in attribute.ConstructorArguments[0].Values)
        {

>             if (constructorArgumentValue.Value is INamedTypeSymbol interfaceSymbol && interfaceSymbol.TypeKind == TypeKind.Interface)

Reviewer: ```suggestion
            if (constructorArgumentValue.Value is INamedTypeSymbol { TypeKind: TypeKind.Interface } interfaceSymbol)
```
